lib/lotus/model/migrator.rb

```diff
@@ -0,0 +1,544 @@
+        duplicated_version_files = files.select { |f| duplicated_versions.include?(f.version) }.group_by { |f| f.version }
+
+        if duplicated_versions.any?
+          message = ['Duplicated versions in following migrations:']
+          duplicated_version_files.each do |vesrion, files|
+            message << "  * #{files.map { |f| f.path }.join(', ')}"
+          end
+          message = message.join("\n")
+
+          raise DuplicatedVersionsError.new(message)
+        end
+      end
+
+      # Returns tuples of migration class, version, and direction
+      #
+      # @since x.x.x
+      # @api private
+      def _get_migration_tuples(current: nil, target: nil)
+        _remove_migration_classes
+        up_mts = []
+        down_mts = []
+        ms = ::Lotus::Model::Migration.descendants
+        applied_migrations = _get_applied_migrations
+
+        files.each do |file|
+          path = file.path
+          version = file.version
+          if target
+            if version > target
+              if applied_migrations.include?(file)
+                load(path)
+                down_mts << [ms.last, version, :down]
+              end
+            elsif !applied_migrations.include?(file)
+              load(path)
+              up_mts << [ms.last, version, :up]
+            end
+          elsif !applied_migrations.include?(file)
+            load(path)
+            up_mts << [ms.last, version, :up]
+          end
+        end
+        up_mts + down_mts.reverse
+      end
+
+      # Returns the dataset for the schema_migrations table. If no such table
+      # exists, it is automatically created.
+      #
+      # @since x.x.x
+      # @api private
+      def schema_dataset
+        c = column
+        ds = db.from(table).order(column)
+
+        if !db.table_exists?(table)
+          db.create_table(table) { String c, primary_key: true }
+        elsif !ds.columns.include?(c)
+          raise(Error, "Migrator table #{table} does not contain column #{c}")
+        end
+        ds
+      end
+
+      # If transactions should be used for the migration, yield to the block
+      # inside a transaction.  Otherwise, just yield to the block.
+      #
+      # @since x.x.x
+      # @api private
+      def _checked_transaction(migration, &block)
+        use_trans = if @use_transactions.nil?
+                      if migration.use_transactions.nil?
+                        @db.supports_transactional_ddl?
+                      else
+                        migration.use_transactions
+                      end
+                    else
+                      @use_transactions
+                    end
+
+        if use_trans
+          db.transaction(&block)
+        else
+          yield
+        end
+      end
+
+      # Remove all migration classes.  Done by the migrator to ensure that
+      # the correct migration classes are picked up.
+      #
+      # @since x.x.x
+      # @api private
+      def _remove_migration_classes
```

Comment:
Sorry for late reply, it is needed because the way Migration aggregate all existing migration is through the `::Lotus::Model::Migration.descendants` which is appended when a class inherits from `Lotus::Model::Migration`.